src/mpid/ch4/shm/xpmem/xpmem_impl.h

```diff
@@ -109,18 +126,12 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_seg_deregist(MPIDI_XPMEM_seg_t * seg)
     int ret;
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_SEG_DEREGIST);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_SEG_DEREGIST);
-
-    /* TODO: decrement segments refcount instead of detach */
-
-    ret = xpmem_detach(seg->vaddr);
-    /* success(0) or failure(-1) */
-    MPIR_ERR_CHKANDJUMP(ret == -1, mpi_errno, MPI_ERR_OTHER, "**xpmem_detach");
-
+    seg->ref_count--;
```

Comment:
Why does it belong to AVL-tree? This is still a segment and you already define a release function for segment. Segments in avl-tree will only be released at the end of the program, I will check `ref_count` there and delete the segment.

Reply:
Releasing a segment and deleting a tree node should be two separate routines. As long as you manage the ref_count in a consistent side, it works for me.

Reply:
I still see you increase the `ref_count` for both `segment` and `segmap` inside the tree-search function, but when you insert a new segment or deregister a segment, you change the `ref_count` in the segment routines. It is inconsistent ! Let's just change `ref_count` only in the segment routines.

Reply:
So my mind is I should increase `ref_count` whenever it is needed. Now I get your point what consistent code means. I believe it is better to increase `ref_count` right away when you actually assign a pointer to the segment. As long as we finally call segment deregister, it would be fine. If you don't agree with this design, I can change the `ref_count` manipulation to all segment function side.

Reply:
Let's keep the `ref_count` changes in the segment function side. It is a parameter belonging to the segment object, not belonging to a tree-node.